### PR TITLE
Code Kingdoms JavaScript Editor

### PR DESCRIPTION
**Pull request by inversion**
This is to add the Code Kingdoms JavaScript editor to TouchDevelop.

I have added some comments in the diff at places which need review.

### CODE REVIEW

File: editor/external.ts [hunk 1]
```diff
@@ -32,7 +32,19 @@ module TDev {
...
-      }, */ {
+      }, */
+        {
+          company: "Code Kingdoms",
+          name: "CK JavaScript",
+          description: "Get down and code JS with the CK editor",
+          id: 'codekingdoms',
+          origin: origin,
+          path: path + 'codekingdoms/editor.html'
+          // Local testing
```
Comment (inversion) [inline editor/external.ts]: do we leave this in? is there a better way for us to get this working locally with the same code?

File: www/codekingdoms/codekingdoms-main.ts [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+    if (st == External.Status.Error)
+      elt.addClass("error");
+    else
+      elt.removeClass("error");
+    box.append(elt);
+    box.scrollTop(box.prop("scrollHeight"));
+  }
+
+  function loadCK(s: string) {
+    // var text = s || "<xml></xml>";
```
Comment (inversion) [inline www/codekingdoms/codekingdoms-main.ts]: review

File: www/codekingdoms/codekingdoms-main.ts [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+    // try {
+    //   Blockly.Xml.domToWorkspace(Blockly.mainWorkspace, xml);
+    // } catch (e) {
+    //   console.error("Cannot load saved Blockly script. Too recent?");
+    //   console.error(e);
+    // }
+  }
+
+  function saveCK(): string {
+    // var xml = Blockly.Xml.workspaceToDom(Blockly.mainWorkspace);
```
Comment (inversion) [inline www/codekingdoms/codekingdoms-main.ts]: review

File: www/codekingdoms/codekingdoms-main.ts [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+        metadata: {
+          name: getName(),
+          comment: getDescription()
+        }
+      },
+    });
+    dirty = false;
+  }
+
+  /*function compileOrError(msgSel?: string) {
```
Comment (inversion) [inline www/codekingdoms/codekingdoms-main.ts]: review dead code

File: www/codekingdoms/codekingdoms-main.ts [hunk 5]
```diff
@@ -0,0 +1,484 @@
...
+      language: External.Language.TouchDevelop,
+      name: getName(),
+    });
+  }*/
+
+  function setupButtons(fota: boolean) {
+    if (fota)
+      $("body").addClass("hasFota");
+
+    // $("#command-quit").click(() => {
```
Comment (inversion) [inline www/codekingdoms/codekingdoms-main.ts]: review dead code

File: www/codekingdoms/editor.html [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            </div>
+            <div>
+              <i id="link-log" class="fa fa-info-circle"></i>
+            </div>
+          </div>
+        </div>
+      </div>
+      <div class="hbox flex1">
+          <div id="editor" class="flex3">
+          		<iframe id="ck_frame" src="http://localhost:8888/ck-client/game/">
```
Comment (inversion) [inline www/codekingdoms/editor.html]: hardcoded localhost

File: www/codekingdoms/style.css [hunk 7]
```diff
@@ -0,0 +1,272 @@
+/* Re-usable, generic rules */
```
Comment (inversion) [inline www/codekingdoms/style.css]: Are all these styles needed or is there duplication from Blockly's CSS?

**Comment by inversion** [comment]
Also noticed the contributing guidelines do not allow merge commits so we can rebase this after fixing the above issues.

File: editor/scriptList.ts [hunk 8]
```diff
@@ -6145,6 +6145,7 @@
 
         static editorIcons = {
             "blockly": "blockeditor,#517EC0",
+            "codekingdoms": "fa-codekingdoms,#fbf299",
```
Comment (pelikhan) [inline editor/scriptList.ts]: is fa-codekingdoms defined in svg.ts?

Comment (inversion) [inline editor/scriptList.ts]: @pelikhan Fixing that now, with regards to the other logo at logoUrl in external.ts I see the format for the Blockly logo at ```https://microbit0.blob.core.windows.net/pub/vtadprog``` is a HTML page containing a PNG. Is this the required format or can we link directly to the PNG?

Comment (inversion) [inline editor/scriptList.ts]: @pelikhan Can you please explain the format for embedding a logo in icons:any in svg.ts? It seems to just be data for one ```<path>```. Can we not embed an arbitrary SVG as our icon?

Comment (pelikhan) [inline editor/scriptList.ts]: You can embed your own svg and I will fix it on our end.

Comment (pelikhan) [inline editor/scriptList.ts]: you could use a data-uri with svg embedded.

**Comment by inversion** [comment]
@pelikhan I've added our SVG logo (small square logo for the editor selection popup) to svg.ts but it's not referenced currently. I have also squashed the commits into one and got rid of the nasty merges.

Thanks for your comments, please let me know what else we need to continue.

**Comment by pelikhan** [comment]
You still need to go through the bot instructions before I can pull this in...
```
In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
```

**Comment by inversion** [comment]
@pelikhan Done, thanks.

**Pull request merged**